pkg/router/controller/host_admitter.go

```diff
@@ -0,0 +1,207 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/watch"
+
+	routeapi "github.com/openshift/origin/pkg/route/api"
+	"github.com/openshift/origin/pkg/router"
+)
+
+// RouteAdmissionFunc determines whether or not to admit a route.
+type RouteAdmissionFunc func(*routeapi.Route) (error, bool)
+
+// SubdomainToRouteMap contains all routes associated with a subdomain -
+// fully qualified and wildcard routes.
+type SubdomainToRouteMap map[string][]*routeapi.Route
+
+// RemoveRoute removes any existing route(s) for a subdomain.
+func (srm SubdomainToRouteMap) RemoveRoute(key string, route *routeapi.Route) bool {
+	k := 0
+	removed := false
+
+	m := srm[key]
+	for i, v := range m {
+		if m[i].Namespace == route.Namespace && m[i].Name == route.Name {
+			removed = true
+		} else {
+			m[k] = v
+			k++
+		}
+	}
+
+	// set the slice length to the final size.
+	m = m[:k]
+
+	if len(m) > 0 {
+		srm[key] = m
+	} else {
+		delete(srm, key)
+	}
+
+	return removed
+}
+
+func (srm SubdomainToRouteMap) InsertRoute(key string, route *routeapi.Route) {
+	// To replace any existing route[s], first we remove all old entries.
+	srm.RemoveRoute(key, route)
+
+	m := srm[key]
+	for idx := range m {
+		if routeapi.RouteLessThan(route, m[idx]) {
+			m = append(m, &routeapi.Route{})
+			// From: https://github.com/golang/go/wiki/SliceTricks
+			copy(m[idx+1:], m[idx:])
+			m[idx] = route
+			srm[key] = m
+
+			// Ensure we return from here as we change the iterator.
+			return
+		}
+	}
+
+	// Newest route or empty slice, add to the end.
+	srm[key] = append(m, route)
+}
+
+// HostAdmitter implements the router.Plugin interface to add admission
+// control checks for routes in template based, backend-agnostic routers.
+type HostAdmitter struct {
+	// plugin is the next plugin in the chain.
+	plugin router.Plugin
+
+	// admitter is a route admission function used to determine whether
+	// or not to admit routes.
+	admitter RouteAdmissionFunc
+
+	// recorder is an interface for indicating route rejections.
+	recorder RejectionRecorder
+
+	// restrictOwnership adds admission checks to restrict ownership
+	// (of subdomains) to a single owner/namespace.
+	restrictOwnership bool
+
+	// subdomainToRoute contains all routes associated with a subdomain
+	// (includes fully qualified and wildcard routes).
+	subdomainToRoute SubdomainToRouteMap
```

Comment:
So this plugin works in conjunction with the `unique_host.go` plugin which does checks for host ownership.  It was only for handling wildcard routes - aah, I can see now that changed because the host no longer has that wildcard spec on it/is no longer returning that boolean indicating it was a wildcard. I'll rework this a bit - we probably just need a map for indicating that - the structure above sounds good. Thx